File: app/src/main/java/com/openmobl/pttDriver/bt/SppSerialSocket.java
package com.openmobl.pttDriver.bt;

import com.openmobl.pttDriver.Constants;

import android.app.Activity;
import android.bluetooth.BluetoothDevice;
import android.bluetooth.BluetoothSocket;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.util.Log;

import java.io.IOException;
import java.security.InvalidParameterException;
import java.util.Arrays;
import java.util.UUID;
import java.util.concurrent.Executors;

public class SppSerialSocket implements Runnable, SerialSocket {
    private static final String TAG = SppSerialSocket.class.getName();

    private static final UUID BLUETOOTH_SPP = UUID.fromString("00001101-0000-1000-8000-00805F9B34FB");

    private final BroadcastReceiver mDisconnectBroadcastReceiver;
    private final BroadcastReceiver mReconnectBroadcastReceiver;

    private final Context mContext;
    private SerialListener mListener;
    private final BluetoothDevice mDevice;
    private BluetoothSocket mSocket;
    private boolean mConnected;

    public SppSerialSocket(Context context, BluetoothDevice device) {
        if(context instanceof Activity)
            throw new InvalidParameterException("expected non UI context");
        mContext = context;
        mDevice = device;
        mDisconnectBroadcastReceiver = new BroadcastReceiver() {
            @Override
            public void onReceive(Context context, Intent intent) {
                disconnect(true);
            }
        };
        mReconnectBroadcastReceiver = new BroadcastReceiver() {
            @Override
            public void onReceive(Context context, Intent intent) {
                SerialListener oldListener = mListener;
                disconnect(true);
                try {
                    connect(oldListener);
                } catch (IOException e) {
                    Log.d(TAG, "reconnect failed");
                    e.printStackTrace();
                }
            }
        };
    }

    @Override
    public String getName() {
        return mDevice.getName() != null ? mDevice.getName() : mDevice.getAddress();
    }

    @Override
    public String getAddress() {
        return mDevice.getAddress();
    }

    private void registerReceivers() {
        mContext.registerReceiver(mDisconnectBroadcastReceiver, new IntentFilter(Constants.INTENT_ACTION_DISCONNECT));
        mContext.registerReceiver(mReconnectBroadcastReceiver, new IntentFilter(Constants.INTENT_ACTION_RECONNECT));
    }

    private void unregisterReceivers() {
        try {
            mContext.unregisterReceiver(mDisconnectBroadcastReceiver);
        } catch (Exception ignored) {
        }
        try {
            mContext.unregisterReceiver(mReconnectBroadcastReceiver);
        } catch (Exception ignored) {
        }
    }

    /**
     * connect-success and most connect-errors are returned asynchronously to listener
     */
    @Override
    public void connect(SerialListener listener) throws IOException {
        if (mConnected)
            throw new IOException("already connected");

        mListener = listener;

        registerReceivers();

        Executors.newSingleThreadExecutor().submit(this);
    }

    @Override
    public void disconnect() {
        disconnect(false);
    }

    @Override
    public void disconnect(boolean silent) {
        //mListener = null; // ignore remaining data and errors
        mConnected = false;
        if(mSocket != null) {
            try {
                mSocket.close();
            } catch (Exception ignored) {
            }
            mSocket = null;
        }

        unregisterReceivers();

        if (!silent && mListener != null)
            mListener.onSerialDisconnect();
        //mListener = null; // ignore remaining data and errors
    }

    @Override
    public void write(byte[] data) throws IOException {
        if (!mConnected)
            throw new IOException("not connected");
        mSocket.getOutputStream().write(data);
    }

    @Override
    public void run() { // connect & read
        Exception sendIoErrorException = null;

        try {
            mSocket = mDevice.createRfcommSocketToServiceRecord(BLUETOOTH_SPP);
            mSocket.connect();

            if(mListener != null)
                mListener.onSerialConnect();
        } catch (Exception e) {
            if(mListener != null)
                mListener.onSerialConnectError(e);

            try {
                mSocket.close();
            } catch (Exception ignored) {
            }
            mSocket = null;
            return;
        }

        mConnected = true;

        try {
            byte[] buffer = new byte[1024];
            int len;

            while (mConnected) {
                len = mSocket.getInputStream().read(buffer);

                byte[] data = Arrays.copyOf(buffer, len);

                if (mListener != null)
                    mListener.onSerialRead(data, null, null);
            }
        } catch (Exception e) {
            Log.d(TAG, "Exception in SPP run loop: " + e);
            e.printStackTrace();

            sendIoErrorException = e;
        }

        mConnected = false;
        if (sendIoErrorException != null && mListener != null)
            mListener.onSerialIoError(sendIoErrorException);
        try {
            mSocket.close();
        } catch (Exception ignored) {
        }
        mSocket = null;
    }

}

File: app/src/main/java/com/openmobl/pttDriver/bt/SppSerialSocket.kt
package com.openmobl.pttDriver.bt

import android.app.Activity
import android.bluetooth.BluetoothDevice
import android.bluetooth.BluetoothSocket
import android.content.*
import android.util.Log
import com.openmobl.pttDriver.Constants
import java.io.IOException
import java.security.InvalidParameterException
import java.util.*
import java.util.concurrent.Executors

class SppSerialSocket(context: Context, device: BluetoothDevice?) : Runnable, SerialSocket {
    private val mDisconnectBroadcastReceiver: BroadcastReceiver
    private val mReconnectBroadcastReceiver: BroadcastReceiver
    private val mContext: Context
    private var mListener: SerialListener? = null
    private val mDevice: BluetoothDevice?
    private var mSocket: BluetoothSocket? = null
    private var mConnected = false

    init {
        if (context is Activity) throw InvalidParameterException("expected non UI context")
        mContext = context
        mDevice = device
        mDisconnectBroadcastReceiver = object : BroadcastReceiver() {
            override fun onReceive(context: Context, intent: Intent) {
                disconnect(true)
            }
        }
        mReconnectBroadcastReceiver = object : BroadcastReceiver() {
            override fun onReceive(context: Context, intent: Intent) {
                val oldListener = mListener
                disconnect(true)
                try {
                    connect(oldListener)
                } catch (e: IOException) {
                    Log.d(TAG, "reconnect failed")
                    e.printStackTrace()
                }
            }
        }
    }

    override val name: String?
        get() = if (mDevice!!.name != null) mDevice.name else mDevice.address
    override val address: String?
        get() = mDevice!!.address

    private fun registerReceivers() {
        mContext.registerReceiver(
            mDisconnectBroadcastReceiver,
            IntentFilter(Constants.INTENT_ACTION_DISCONNECT)
        )
        mContext.registerReceiver(
            mReconnectBroadcastReceiver,
            IntentFilter(Constants.INTENT_ACTION_RECONNECT)
        )
    }

    private fun unregisterReceivers() {
        try {
            mContext.unregisterReceiver(mDisconnectBroadcastReceiver)
        } catch (ignored: Exception) {
        }
        try {
            mContext.unregisterReceiver(mReconnectBroadcastReceiver)
        } catch (ignored: Exception) {
        }
    }

    /**
     * connect-success and most connect-errors are returned asynchronously to listener
     */
    @Throws(IOException::class)
    override fun connect(listener: SerialListener?) {
        if (mConnected) throw IOException("already connected")
        mListener = listener
        registerReceivers()
        Executors.newSingleThreadExecutor().submit(this)
    }

    override fun disconnect() {
        disconnect(false)
    }

    override fun disconnect(silent: Boolean) {
        //mListener = null; // ignore remaining data and errors
        mConnected = false
        if (mSocket != null) {
            try {
                mSocket!!.close()
            } catch (ignored: Exception) {
            }
            mSocket = null
        }
        unregisterReceivers()
        if (!silent && mListener != null) mListener!!.onSerialDisconnect()
        //mListener = null; // ignore remaining data and errors
    }

    @Throws(IOException::class)
    override fun write(data: ByteArray) {
        if (!mConnected) throw IOException("not connected")
        mSocket!!.outputStream.write(data)
    }

    override fun run() { // connect & read
        var sendIoErrorException: Exception? = null
        try {
            mSocket = mDevice!!.createRfcommSocketToServiceRecord(BLUETOOTH_SPP)
            mSocket.connect()
            if (mListener != null) mListener!!.onSerialConnect()
        } catch (e: Exception) {
            if (mListener != null) mListener!!.onSerialConnectError(e)
            try {
                mSocket!!.close()
            } catch (ignored: Exception) {
            }
            mSocket = null
            return
        }
        mConnected = true
        try {
            val buffer = ByteArray(1024)
            var len: Int
            while (mConnected) {
                len = mSocket.getInputStream().read(buffer)
                val data = Arrays.copyOf(buffer, len)
                if (mListener != null) mListener!!.onSerialRead(data, null, null)
            }
        } catch (e: Exception) {
            Log.d(TAG, "Exception in SPP run loop: $e")
            e.printStackTrace()
            sendIoErrorException = e
        }
        mConnected = false
        if (sendIoErrorException != null && mListener != null) mListener!!.onSerialIoError(
            sendIoErrorException
        )
        try {
            mSocket.close()
        } catch (ignored: Exception) {
        }
        mSocket = null
    }

    companion object {
        private val TAG = SppSerialSocket::class.java.name
        private val BLUETOOTH_SPP = UUID.fromString("00001101-0000-1000-8000-00805F9B34FB")
    }
}